crates/katana/primitives/src/transaction.rs

```diff
@@ -241,33 +377,109 @@ impl L1HandlerTx {
+    ///
+    /// Used as a simple replay attack protection.
     pub chain_id: ChainId,
-    pub version: FieldElement,
-    pub class_hash: ClassHash,
+    /// The nonce value of the account. Corresponds to the number of transactions initiated by
+    /// sender.
+    pub nonce: Nonce,
+    /// The transaction signature associated with the sender address.
     pub signature: Vec<FieldElement>,
+    /// The hash of the contract class from which the account contract will be deployed from.
+    pub class_hash: ClassHash,
+    /// The contract address of the account contract that will be deployed.
     pub contract_address: ContractAddress,
+    /// The salt used to generate the contract address.
     pub contract_address_salt: FieldElement,
+    /// The input data to the constructor function of the contract class.
     pub constructor_calldata: Vec<FieldElement>,
+    /// Resource bounds for the transaction execution
+    pub resource_bounds: ResourceBoundsMapping,
+    /// The tip for the transaction
+    pub tip: u64,
+    /// Data needed to allow the paymaster to pay for the transaction in native tokens
+    pub paymaster_data: Vec<FieldElement>,
+    /// The storage domain of the account's nonce (an account has a nonce per da mode)
+    pub nonce_data_availability_mode: DataAvailabilityMode,
+    /// The storage domain of the account's balance from which fee will be charged
+    pub fee_data_availability_mode: DataAvailabilityMode,
 }
 
 impl DeployAccountTx {
     /// Compute the hash of the transaction.
     pub fn calculate_hash(&self, is_query: bool) -> TxHash {
-        compute_deploy_account_v1_tx_hash(
-            self.contract_address.into(),
-            self.constructor_calldata.as_slice(),
-            self.class_hash,
-            self.contract_address_salt,
-            self.max_fee,
-            self.chain_id.into(),
-            self.nonce,
-            is_query,
-        )
+        match self {
+            DeployAccountTx::V1(tx) => compute_deploy_account_v1_tx_hash(
+                tx.contract_address.into(),
+                &tx.constructor_calldata,
+                tx.class_hash,
+                tx.contract_address_salt,
+                tx.max_fee,
+                tx.chain_id.into(),
+                tx.nonce,
+                is_query,
+            ),
+
+            DeployAccountTx::V3(tx) => compute_deploy_account_v3_tx_hash(
+                tx.contract_address.into(),
+                &tx.constructor_calldata,
+                tx.class_hash,
+                tx.contract_address_salt,
+                tx.tip,
+                &tx.resource_bounds.l1_gas,
+                &tx.resource_bounds.l2_gas,
+                &tx.paymaster_data,
+                tx.chain_id.into(),
+                tx.nonce,
+                &tx.nonce_data_availability_mode,
+                &tx.fee_data_availability_mode,
+                is_query,
+            ),
+        }
```

Comment:
You prefer to keep the function with all arguments to avoid too much coupling between tx type and actual hash computation? We could impl a hasher trait or something similar on the TX itself to have perhaps something more expressive?

Reply:
May be a good first issue actually.
+ adding the links to the starknet docs.